README.md
# Pelican-Twitter Bridge

This set of Python scripts are aimed at allowing users of the [Pelican](http://getpelican.com) blog engine to link some of their posts to Twitter.

## Presentation

For now, I only wrote two scripts:

- `pelican_auto_tweet.py`
- `pelican_tweet_summary.py`

The first one is usefull to automatically post a tweet on Twitter for the latest blog post. As Pelican is text-based and often used with Git, this script makes use of Git logs to find new posts. More details are given below.

The second script posts a tweet on Twitter for each article posted on the blog in the last 7 days. Like the previous script, I give some details below.

## Installation

Just fork this repo! :)

Since I use two external libraries, you also need to install them if you want to use these scripts.

### Python-Twitter

Located at [http://code.google.com/p/python-twitter/](http://code.google.com/p/python-twitter/), the steps for installing it are : 

```bash
hg clone http://python-twitter.googlecode.com/hg/ python-twitter
cd python-twitter
hg update
python setup.py build
python setup.py install
```

### Bitly 

To shorten links, I use Bitly. This API is available in [Pip repository](http://www.pip-installer.org/en/latest/installing.html "How to install Pip"), so installation is quite easy.

```bash
pip install bitly_api
```

The results of requests made to Bitly API are Unicode encoded, so I also need a library to re-encode them. This library is [Unidecode](https://pypi.python.org/pypi/Unidecode). _If you have a better way to work with Bitly, I will appreciate if you can tell me about!_

```Python
git clone http://www.tablix.org/~avian/git/unidecode.git unidecode
python setup.py build
python setup.py install
```

## Usage

The script need some configuration. For this, use the file `conf.py`.

```bash
/
|- conf.py
|- pelican_auto_tweet.py
|- pelican_tweet_summary.py
|- README.md
```

In this script, you **have** to define four variables for the [Twitter API](https://dev.twitter.com): 

```Python
CONSUMER_KEY = ''
CONSUMER_SECRET = ''
ACCESS_TOKEN = ''
ACCESS_TOKEN_SECRET = ''
```

And two variables for the [Bitly API](https://bitly.com/). _For now, Bitly is required in `pelican_auto_tweet.py`, but I plan to make it optional. It's already optional for `pelican_tweet_summary.py`_

```Python
BITLY_USER = ''
BITLY_API_KEY = ''
```

You can also define in a variable the base directory of you Pelican blog. This option is optional.

```Python
BASE_DIR = '~/pelican_blog/'
```

Now, you can use the scripts.

### Pelican_auto_tweet

I use this script to tweet the link of my **last** blog post. 

#### TL;DR; 

This script do 4 things :

1. It checks if a new article was writen (by looking if the [Git [FR]](http://blog.quack1.me/tag/git.html "Blog Quack1 - Tag « Git »") _commit_ log starts with `[POST]`).
2. In this case, it pushes the new _commits_ on the [default git repository [FR]](http://blog.quack1.me/git_push_multiple_remote.md "Git : Pusher ses modifications sur plusieurs dépôts en une seule commande").
3. It updates the blog on the serveur through SSH (command `make ssh_upload` for the _Pelican-ists_).
4. It posts a tweet.

#### Long explanation

I use Git to backup my blog. After each new blog post, I commit the new file with a commit message that starts with `[POST]`. I adapted the script to my own needs, so it get the last commit message, and if the message starts with `[POST]`, it tweets it. 

Before posting a tweet, the script pushes the new commits to the default Git repository, and updates the blog using Pelican Makefile.

The script gets the title and the URL of the lasts posts directly from the source file. If there are many files involved in the commit, every file that is located in `content/` and has a `.md` or `.rst` extension will be considered as one of the lasts posts for which a tweet will be send.

**Tweet text**

The tweet message is constructed based on a format given by the user.

This configuration is set in `conf.py`, with the variable `TWEET_FORMAT_AUTO`. It's simply a string that contains variables, that will be replaced by real values during execution of the script. 

Available variables are : 

- `$$POST_TITLE$$` : Will be replaced by the title of the article.
- `$$POST_URL$$` : Will be replaced by the URL of the article.

If the variable `TWEET_FORMAT_AUTO` is not defined, the default one will be used : 

```Python
TWEET_FORMAT_AUTO = '$$POST_TITLE$$ $$POST_URL$$ #blog'
```

The post title and URL come from the post source file. `$$POST_TITLE$$` is the title of the post (post variable `Title:`). To construct the URL, the script extract the URL of the site in the `pelicanconf.py` file (variable `SITEURL=`), and append to it the slug of the blog post (variable `Slug:` in the header of the blog post file).


#### Execution

To launch the script, you have some options.

1. Run it directly from the root of your Pelican directory:
	```
	$ python ~/pelican_auto_tweet.py
	```
2. Run it from anywhere, and give it the path to the root of your Pelican directory:
	```
	$ python ~/pelican_auto_tweet.py ~/pelican_blog/
	```	



### Pelican_tweet_summary

This script sends one tweet for each blog post published in the last few days.

I use the same method I used for the previous script to get the URL of the posts, or their title.


#### Number of days for which posts should be considered as new

By default, the value is 7 days.

This value can be modified in `conf.py` (here, 14 days):

```Python
SUMMARY_DAYS = 14
```


#### Time between two Twitter updates

By default, each tweet is sent 3 minutes (180seconds) after the previous one.

It's possible to use a different interval using a variable in the configuration file `conf.py` (here, 60 seconds): 

```Python
SUMMARY_INTERVAL = 60
```


#### Tweet text

3 types of tweets are sent : 

1. `TWEET_SUMMARY_BEGIN` : A message to inform that the « blog replay » is starting.
2. `TWEET_FORMAT_SUMMARY` : A message for each post.
3. `TWEET_SUMMARY_END_ONE` and `TWEET_SUMMARY_END_MANY` : A message to inform that the « blog replay » is finished.

The first tweet can (or not) contains one variable : the name/url of the website. For this, use `$$BLOG_URL$$`. A second variable is mandatory. Place `%d` anywhere you want, and this variable will be replaced by the number of days for which we check for new posts. Exemple : `Here are the articles posts in the last %d days on $$BLOG_URL$$` will become `Here are the articles posts in the last 7 days on http://blog.quack1.me`.
If you don't want to use send a first tweet, just set the `IS_TWEET_SUMMARY_BEGIN` variable to `False` in the configuration file :

```Python
IS_TWEET_SUMMARY_BEGIN = False
```

The second tweet can contains variables like for `pelican_auto_tweet.py` (see above).

There are two cases for the last tweet : 

1. Only one new post was posted. The `TWEET_SUMMARY_END_ONE` is sent (no variable allowed).
2. Many posts were posted. The `TWEET_SUMMARY_END_MANY` is sent. One variable is mandatory. Add `%d` in your text. This will be replaced by the numbered of new posts. Exemple : `There were %d articles` will become `There were 24 articles`.

As for the first tweet, you can define the last tweet not to be sent. Set the `IS_TWEET_SUMMARY_END` variable to `False` in the configuration file :

```Python
IS_TWEET_SUMMARY_END = False
```

&nbsp;

All these 4 variables can be ommited, and will be replaced by default values : 

```Python
TWEET_FORMAT_SUMMARY   = '#blogReplay $$POST_TITLE$$ $$POST_URL$$ #blog'
IS_TWEET_SUMMARY_BEGIN = True
TWEET_SUMMARY_BEGIN    = '#blogReplay Voici les articles publiés ces 7 derniers jours sur $$BLOG_URL$$'
TWEET_SUMMARY_END      = True
TWEET_SUMMARY_END_ONE  = '#blogReplay C'était le seul article publié cette semaine. Fin du spam! :)'
TWEET_SUMMARY_END_MANY = '#blogReplay C'était les %d articles publiés cette semaine. Fin du spam! :)'
```

#### Execution

The usage of this script is quite the same as the previous one

1. Run it directly from the root of your Pelican directory:
	```
	$ python ~/pelican_tweet_summary.py
	```
2. Run it from anywhere, and give it the path to the root of your Pelican directory:
	```
	$ python ~/pelican_tweet_summary.py ~/pelican_blog/
	```

## More

I wrote some posts on my blog about this script : 

- [Presentation](http://blog.quack1.me/pelican_auto_tweet-en.html) ;
- [How to automate the publication with Git hooks](http://blog.quack1.me/git_hooks_pelican-en.html).

I also created the theme I use on my blog, which is available [on Github](https://github.com/quack1/notebook).


CHANGELOG
CHANGELOG
=========

v0.9
----

- Add a Pelican_get_posts script to display all the posts, orderd by asc. date



libpelican.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

'''A library that provides usefull functions to interact with Pelican
blog posts.
'''

__author__     = 'quack1'
__version__    = '0.9'
__date__       = '2014-05-27'
__copyright__  = 'Copyright © 2013-2014, Quack1'
__licence__    = 'BSD'
__credits__    = ['Quack1']
__maintainer__ = 'Quack1'
__email__      = 'quack+pelican@quack1.me'
__status__     = 'Development'

import datetime
import os.path
import re

class PelicanBlog:
	'''Represents a Pelican blog.

	This class gives a clean an simple API to interact with a Pelican blog.
	
	After being created with the base directory of the blog, functions allows
	to obtain information about an article.

	The information that's possible to get for an article are : 
	  - author
	  - category
	  - datetime
	  - lang
	  - slug
	  - status
	  - summary
	  - tags
	  - title
	  - url
	
	It is also possible to get information from the whole blog, like : 
	  - A list of all posts
	  - The base URL
	  - The base directory
	  - All the drafts.
	'''


	_blog_directory    = None
	_content_directory = None
	_blog_base_url     = None

	_REGEX_BASE_URL = re.compile(r'SITEURL = \'(.*)\'',re.IGNORECASE)

	def __init__(self, base_directory):
		'''Initiate a new Pelican blog instance.

		Args:
		  base_directory:
		    The directory in which the blog stands.
		    This directory **has to** contain the `content` directory 
		    and the `pelicanconf.py` and `Makefile` files.
		'''
		self._blog_directory = base_directory
		self._content_directory = os.path.join(self._blog_directory, 'content')

	def get_site_base_url(self):
		'''Get the blog base URL from the `pelicanconf.py` file.
		The file must be located in the base directory given in the parameter 
		of the constructor.

		The file which will be read will be {base_directory}/pelicanconf.py.
		'''
		if self._blog_base_url:
			return self._blog_base_url
		else:
			with open(os.path.join(self._blog_directory,"pelicanconf.py")) as f:
				for l in f:
					res = self._REGEX_BASE_URL.search(l)
					if res:
						self._blog_base_url = res.group(1)
						if not self._blog_base_url.endswith('/'):
							self._blog_base_url += '/'
						return self._blog_base_url


	def _get_post_info(self, post_filename, post_info):
		'''Get the value of one option field in a post file.
		The file must be located in a `content` directory, located in the base
		directory given in the parameter of the constructor.

		The file which will be read will be `{base_directory}/content/{post_filename}`.

		Args:
		  post_filename:
		    The filename of the article.
		  post_info:
		    The name of the wanted information.
		    The value returned will be get using this regular expression:
		      `{post_info}\s*:\s*(.*)$`
		      If `post_info` is 'Title', the regular expression will match
		      on the following line : "Title: My Very Good Blog Post".
		      The returned value will be "My Very Good Blog Post".
		'''
		regex = re.compile("^%s\s*:\s*(.*)$"%post_info, re.IGNORECASE)

		with open(os.path.join(self._content_directory, post_filename),"r") as f:
			for line in f:
				res = regex.search(line)
				if res:
					return(res.group(1))
			return ""

	def get_base_directory(self):
		'''Get the base directory of a blog instance.

		Returns:
		  The base directory containing a blog instance
		'''
		return self._blog_directory

	def get_post_title(self, post_filename):
		'''Get the title of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The title of the article, or a blank string.
		'''
		return self._get_post_info(post_filename, "Title")

	def get_post_date(self, post_filename):
		'''Get the date of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The date of the article in a **datetime** object.
		  If the date is not defined in the article, None is returned.
		'''
		date = self._get_post_info(post_filename, "Date").strip()
		if date is None or date == "":
			return None
		else:
			return datetime.datetime.strptime(date, "%Y-%m-%d %H:%M")

	def get_post_author(self, post_filename):
		'''Get the author of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The author of the article, or a blank string.
		'''
		return self._get_post_info(post_filename, "Author")

	def get_post_category(self, post_filename):
		'''Get the category of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The category of the article, or a blank string.
		'''
		return self._get_post_info(post_filename, "Category")

	def get_post_tags(self, post_filename):
		'''Get the tags of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  A list containing all the tags of the article. If no tags is defined, the
		    list can contain only an empty string.

		TODO: If the string returned by get_post_info is blank, return an empty
		  list.
		'''
		tags = self._get_post_info(post_filename, "Tags").split(',')
		return [x.strip() for x in tags]

	def get_post_slug(self, post_filename):
		'''Get the slug of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The slug of the article, or a blank string.

		TODO: If the slug option is not defined, generate the URL with
		  the Pelican core method.
		'''
		return self._get_post_info(post_filename, "Slug")

	def get_post_summary(self, post_filename):
		'''Get the summary of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The summary of the article, or a blank string.
		'''
		return self._get_post_info(post_filename, "Summary")

	def get_post_lang(self, post_filename):
		'''Get the lang of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The lang of the article, or a blank string.
		'''
		return self._get_post_info(post_filename, "Lang")

	def get_post_status(self, post_filename):
		'''Get the status of a blog post.
		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The status of the article, or a blank string.
		'''
		return self._get_post_info(post_filename, "Status")

	def is_draft(self, post_filename):
		'''Returns the 'draft' status of a blog post.
		If the uppercase status of the blog post is 'DRAFT', it is considered
		as a draft.

		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  True if the blog post is a draft.
		'''
		return self.get_post_status(post_filename).upper() == 'DRAFT'

	def posts_have_drafts(self, posts_filenames):
		'''Look if a list of articles contains drafts.
		Args:
		  posts_filenames:
		    A list containing the filename of the articles.

		Returns:
		  True if at least one article in the list is a draft.
		'''
		for f in posts_filenames:
			if self.get_post_status(f).upper() == 'DRAFT':
				return True
		return False

	def get_post_url(self, post_filename):
		'''Get the complete URL of a blog post.
		The URL is generated by appending the post slug after the site 
		base URL.

		Args:
		  post_filename:
		    The filename of the article.

		Returns:
		  The complete URL to the article.
		'''
		return self.get_site_base_url() + self.get_post_slug(post_filename) + ".html"

	def get_posts(self):
		'''Get the list of all articles of the blog.

		Returns:
		  A list containing all the articles of the blog.
		'''
		l = os.listdir(self._content_directory)
		posts = []
		for post_filename in l:
			base,ext = os.path.splitext(post_filename)
			if ext in ('.rst','.md'):
				posts.append(post_filename)
		return posts
	
	def get_drafts(self):
		'''Get the list of all drafts of the blog.

		To be considered as a draft, the status of the article (get with the
			`get_post_status()` method), uppercase, has to be 'DRAFT'.

		Returns:
		  A list containing all the drafts of the blog.
		'''
		l = os.listdir(self._content_directory)
		posts = []
		for post_filename in l:
			base,ext = os.path.splitext(post_filename)
			if ext in ('.rst','.md'):
				if self.get_post_status(post_filename).upper() == "DRAFT":
					posts.append(post_filename)
		return posts


conf_sample.py
#!/usr/bin/env python
# -*- coding: utf-8 -*- #

'''This file is part of the pelican_auto_tweet project. 
It provides configuration variables used into the different scrips of the project.
'''

__author__     = 'quack1'
__version__    = '0.9'
__date__       = '2014-05-27'
__copyright__  = 'Copyright © 2013-2014, Quack1'
__licence__    = 'BSD'
__credits__    = ['Quack1']
__maintainer__ = 'Quack1'
__email__      = 'quack@quack1.me'
__status__     = 'Development'

import os.path

class Global:
	blog_directory = '<Pelican_Blog_Home_Directory>'


class Auto:
	tweet_format = '$$POST_TITLE$$ $$POST_URL$$ #blog'

class Summary:
	days     = 10
	interval = 180

	# Twitter conf
	tweet_format          = '#blogReplay $$POST_TITLE$$ $$POST_URL$$ #blog'
	tweet_format_begin    = '#blogReplay Voici les articles publiés ces 7 derniers jours sur $$BLOG_URL$$'
	tweet_format_end_one  = '#blogReplay C\'était le seul article publié cette semaine. Fin du spam! :)'
	tweet_format_end_many = '#blogReplay C\'était les %d articles publiés cette semaine. Fin du spam! :)'
	tweet_begin           = True
	tweet_end             = False


class Links:
	twitter_username = ""
	page_header      = """Title: Liens partagés sur Twitter
	Date: %s
	Author: Quack1"""
	out_file = os.path.join(Global.blog_directory, "content/pages/links.md")


class Twitter:
	consumer_key        = ''
	consumer_secret     = ''
	access_token        = ''
	access_token_secret = ''


class Bitly:
	user    = ""
	api_key = ""


pelican_auto_tweet.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

'''This script is used to automaticaly tweet link of the last published post from
a Pelican blog.
Both Markdown and Rest syntax are supported. The 'Slug' tag **has** to be set
	so the script can create the complete URL of the post.
'''

__author__     = 'quack1'
__version__    = '0.9'
__date__       = '2014-05-27'
__copyright__  = 'Copyright © 2013-2014, Quack1'
__licence__    = 'BSD'
__credits__    = ['Quack1']
__maintainer__ = 'Quack1'
__email__      = 'quack@quack1.me'
__status__     = 'Development'

import re
import os.path
import sys
import commands
import unicodedata
import os
from HTMLParser import HTMLParser

# From https://github.com/bitly/bitly-api-python
import bitly_api as bitlyapi
# From http://code.google.com/p/python-twitter/
import twitter

import conf
import libpelican

TWITTER_API = None
BITLY_API   = None
BLOG        = None

#
# From http://stackoverflow.com/questions/753052/strip-html-from-strings-in-python
#
class MLStripper(HTMLParser):
    def __init__(self):
        self.reset()
        self.fed = []
    def handle_data(self, d):
        self.fed.append(d)
    def get_data(self):
        return ''.join(self.fed)
def strip_tags(html):
    s = MLStripper()
    s.feed(html)
    return s.get_data()

def twitter_connect():
	'''Connect the API to Twitter with the OAuth protocol.'''
	global TWITTER_API
	TWITTER_API = twitter.Api(consumer_key=conf.Twitter.consumer_key, 
		consumer_secret=conf.Twitter.consumer_secret, access_token_key=conf.Twitter.access_token_key, 
    	access_token_secret=conf.Twitter.access_token_secret)

def twitter_send(text):
	'''Post a new tweet on Twitter.
	If the API is not connected, a call to twitter_connect() is made.'''
	if not TWITTER_API:
		twitter_connect()
	TWITTER_API.PostUpdate(text)
	print(text)


# Check arguments
# If one argument is passed, it will be considered as the base blog directory.
# If it's not present, the current working directory will be used.
if len(sys.argv) >= 2:
	base_dir = sys.argv[1]
elif conf.Global.blog_directory:
	base_dir = conf.Global.blog_directory
else:
	base_dir = './'

print base_dir

# A new blog instance is created on this directory.
BLOG = libpelican.PelicanBlog(base_dir)

# Try to connect to BitlyAPI
try:
	if not conf.Bitly.user:
		print("Error. No conf.Bitly.user defined.")
	if not conf.Bitly.api_key:
		print("Error. No conf.Bitly.api_key defined.")
	BITLY_API = bitlyapi.Connection(conf.Bitly.user, conf.Bitly.api_key)
	# Test Bitly connection. If the link to Bitly API is down, an exception is thrown and
	# Bitly will not be used.
	BITLY_API.lookup(BLOG.get_site_base_url())
except:
	BITLY_API = None

# Move to the blog base directory, and get the last commit message and the last
# commited files. 
os.chdir(BLOG.get_base_directory())
result = commands.getoutput('git show --pretty="format:%s" --name-only')

# The first line is the commit message. The following lines contain the filenames.
log_message = result.splitlines()[0]
files       = result.splitlines()[1:]

# Get the tweet format from the configuration file. If the format is not present,
# a default one is created.
if not conf.Auto.tweet_format:
	TWEET_FORMAT_AUTO = '$$POST_TITLE$$ $$POST_URL$$ #blog'
else:
	TWEET_FORMAT_AUTO = conf.Auto.tweet_format

# If the log message starts by '[POST]', it means that (at least) one new post was 
# writen and needs to be published. 
if log_message.startswith('[POST]'):

	f = []
	for x in files:
		name = os.path.basename(x)
		b,ext = os.path.splitext(name)
		if ext in ('.md', '.rst'):
			f.append(name)
	# Check that there is no drafts in the commited articles. 
	if BLOG.posts_have_drafts(f):
		print "You are about to publish drafts."
		response = raw_input("Do you want to publish them (y-n) ? [n]")
		if not response:
			response = "N"
		while response.upper() not in ("N", "Y"):
			response = raw_input("Do you want to publish them (y-n) ? [n]")
			if not response:
				response = "N"
		if response.upper() == "N":
			sys.exit(2)

	# Pull the last modifications in the git repository
	os.system('git pull --commit --no-edit')
	# Push our updates
	os.system('git push')
	# Generate and upload the blog
	os.system('make ssh_upload')

	# A new tweet is sent for each commited article.
	for filename in files:
		post_filename = os.path.basename(filename)
		base, ext     = os.path.splitext(filename)
		# An article is a '.rst' or '.md' file in the 'content/' directory.
		if base.startswith('content/') and ext in ('.rst','.md'):
			title = BLOG.get_post_title(post_filename)
			url   = BLOG.get_post_url(post_filename)

			# Shorten the link if Bitly is used.
			if BITLY_API:
				s = BITLY_API.shorten(url)
				u = s['url']
				url = unicodedata.normalize('NFKD', u).encode('ascii','ignore')

			# Generate tweet message from TWEET_FORMAT_AUTO
			tweet_text = TWEET_FORMAT_AUTO.replace('$$POST_TITLE$$', title)
			tweet_text = tweet_text.replace('$$POST_URL$$', url)
			tweet_text = strip_tags(tweet_text)

			# Post tweet
			twitter_send(tweet_text)

#END


pelican_get_drafts.py
#! /usr/bin/env python
# -*- coding: utf-8 -*- 

'''This script prints the drafts present on a Pelican blog.
If no argument is passed, the current directory ('.') is used.
'''

__author__     = 'quack1'
__version__    = '0.9'
__date__       = '2014-05-21'
__copyright__  = 'Copyright © 2013-2014, Quack1'
__licence__    = 'BSD'
__credits__    = ['Quack1']
__maintainer__ = 'Quack1'
__email__      = 'quack@quack1.me'
__status__     = 'Development'

import libpelican
import sys
import os.path
import datetime

# Check arguments
# If one argument is passed, it will be considered as the base blog directory.
# If it's not present, the current working directory will be used.
if len(sys.argv) >= 2:
	base_dir = sys.argv[1]
else:
	base_dir = './'

# A new blog instance is created on this directory.
BLOG = libpelican.PelicanBlog(base_dir)

# Get the drafts from the blog instance and print them.
drafts = BLOG.get_drafts()
for article in drafts:
	print "File '%s' ([%s] %s)"%(article, BLOG.get_post_date(article), BLOG.get_post_title(article))

# END


pelican_get_articles.py
#! /usr/bin/env python
# -*- coding: utf-8 -*- 

'''This script prints the articles present on a Pelican blog ordered by post date asc.
If no argument is passed, the current directory ('.') is used.
'''

__author__     = 'quack1'
__version__    = '0.9'
__date__       = '2014-07-07'
__copyright__  = 'Copyright © 2013-2014, Quack1'
__licence__    = 'BSD'
__credits__    = ['Quack1']
__maintainer__ = 'Quack1'
__email__      = 'quack@quack1.me'
__status__     = 'Development'

import libpelican
import sys
import os.path
import datetime

# Check arguments
# If one argument is passed, it will be considered as the base blog directory.
# If it's not present, the current working directory will be used.
if len(sys.argv) >= 2:
	base_dir = sys.argv[1]
else:
	base_dir = './'

# A new blog instance is created on this directory.
BLOG = libpelican.PelicanBlog(base_dir)

# Get the drafts from the blog instance and print them.
posts = BLOG.get_posts()
posts = sorted(posts, key=lambda x: BLOG.get_post_date(x))
for article in posts:
	print "File '%s' ([%s] %s)"%(article, BLOG.get_post_date(article), BLOG.get_post_title(article))

# END


pelican_tweet_summary.py
#! /usr/bin/env python
# -*- coding: utf-8 -*- 

'''This script is used to automaticaly tweet links to the blog posts published in the
	last 7 days on the blog.
Both Markdown and Rest syntax are supported. The 'Slug' tag **has** to be set
	so the script can create the complete URI to the post.
'''

__author__     = 'quack1'
__version__    = '0.9'
__date__       = '2014-05-27'
__copyright__  = 'Copyright © 2013-2014, Quack1'
__licence__    = 'BSD'
__credits__    = ['Quack1']
__maintainer__ = 'Quack1'
__email__      = 'quack@quack1.me'
__status__     = 'Development'

import datetime
import os
import re
import sys
import time
from HTMLParser import HTMLParser

# From http://code.google.com/p/python-twitter/
import twitter
# From https://github.com/bitly/bitly-api-python
import bitly_api as bitlyapi
import unicodedata

import conf
import libpelican



TODAY         = datetime.date.today()
TWITTER_API   = None
FIRST_TWEETED = False
SLUGS         = []
POSTS         = []
BITLY_API     = None
LOG_FILE      = '/var/log/pelican_auto_tweet/summary.log'

#
# From http://stackoverflow.com/questions/753052/strip-html-from-strings-in-python
#
class MLStripper(HTMLParser):
    def __init__(self):
        self.reset()
        self.fed = []
    def handle_data(self, d):
        self.fed.append(d)
    def get_data(self):
        return ''.join(self.fed)
def strip_tags(html):
    s = MLStripper()
    s.feed(html)
    return s.get_data()


def twitter_connect():
	'''Connect the API to Twitter with the OAuth protocol.'''
	global TWITTER_API
	TWITTER_API = twitter.Api(consumer_key=conf.Twitter.consumer_key, 
		consumer_secret=conf.Twitter.consumer_secret, access_token_key=conf.Twitter.access_token_key, 
		access_token_secret=conf.Twitter.access_token_secret)

def twitter_send(text):
	'''Post an update on Twitter.
	If the API is not connected, a call to twitter_connect() is made.'''
	if not TWITTER_API:
		twitter_connect()
	TWITTER_API.PostUpdate(text)

def first_tweet(msg):
	global FIRST_TWEETED
	if not FIRST_TWEETED:
		twitter_send(msg)
		FIRST_TWEETED = True

# Check arguments
# If one argument is passed, it will be considered as the base blog directory.
# If it's not present, the current working directory will be used.
if len(sys.argv) >= 2:
	base_dir = sys.argv[1]
elif conf.Global.blog_directory:	# BASE_DIR comes from conf.py
	base_dir = conf.Global.blog_directory
else:
	base_dir = './'

# A new blog instance is created on this directory.
BLOG = libpelican.PelicanBlog(base_dir)

# Get in the configuration file the number of days for which posts should be 
# considered as new. If the configuration file is not present a default value
# is used.
days = 0
try:
	if conf.Summary.days < 0:
		days = 0
	else:
		days = conf.Summary.days
except:
	print('Using default number of days : 7')
	days = 7

# Get in the configuration file the interval between two new twitter updates.  
# If the configuration file is not present a default value is used. The value 
# is in seconds.
interval = 0
try:
	if conf.Summary.interval < 0:
		interval = 0
	else:
		interval = conf.Summary.interval
except:
	print('Using default number of interval : 180s')
	interval = 180


# Trying to connect to BitlyAPI
try:
	if not conf.Bitly.user:
		print("Error. No conf.Bitly.user defined.")
	if not conf.Bitly.api_key:
		print("Error. No conf.Bitly.api_key defined.")
	BITLY_API = bitlyapi.Connection(conf.Bitly.user, conf.Bitly.api_key)
	# Test Bitly connection. If the link to Bitly API is down, an exception is 
	# thrown and Bitly will not be used.
	BITLY_API.lookup(BLOG.get_site_base_url())
except:
	BITLY_API = None


# Get all the blog articles. For every article published in the last 
# `conf.Summary.days` days, a tweet will be published.
# Tweets are not sent for the drafts.
# A new array is created with the blog posts that have to be tweeted.
# This array contains dicts containing :
#	- the URL of the article
#	- the date of the article
#	- and its title.
# This array is finally sorted by the publication date.
for post_filename in BLOG.get_posts():
	if not BLOG.is_draft(post_filename):
		post_date = BLOG.get_post_date(post_filename).date()
		if post_date + datetime.timedelta(days) >= TODAY:
			title = BLOG.get_post_title(post_filename)
			url   = BLOG.get_post_url(post_filename)
			if not url in SLUGS:
				p = dict()
				p['url'] = url
				p['date'] = post_date
				p['title'] = title
				SLUGS.append(url)
				POSTS.append(p)
POSTS.sort(key=lambda item:item['date'])

# Get the tweet format from the configuration file. If the format is not present,
# a default one is created.
if not conf.Summary.tweet_format:
	TWEET_FORMAT = '#blogReplay $$POST_TITLE$$ $$POST_URL$$ #blog'
else:
	TWEET_FORMAT = conf.Summary.tweet_format

# Get the length of 'pure' text of the TWEET_FORMAT. 
# We suppress, by now, two variables : 
#	- $$POST_TITLE$$ -> Will be replaced by the title of the article
#	- $$POST_URL$$ -> Will be replaced by the URL to the article
TWEET_FORMAT_LENGTH = len(TWEET_FORMAT)
if '$$POST_TITLE$$' in TWEET_FORMAT:
	TWEET_FORMAT_LENGTH -= len('$$POST_TITLE$$')
if '$$POST_URL$$' in TWEET_FORMAT:
	TWEET_FORMAT_LENGTH -= len('$$POST_URL$$')

# Get the text for the first tweet from the configuration file. If the format is 
# not present, a default one is used.
# Only one variable can be used in the format : 
#	- $$BLOG_URL$$ -> Will be replaced by the URL ot the blog
if conf.Summary.tweet_begin:
	if not conf.Summary.tweet_format_begin:
		TWEET_BEGIN = '#blogReplay Voici les articles publiés ces %d derniers jours sur $$BLOG_URL$$'%(days)
	else:
		TWEET_BEGIN = conf.Summary.tweet_format_begin
	if '$$BLOG_URL$$' in TWEET_BEGIN:
		TWEET_BEGIN = TWEET_BEGIN.replace('$$BLOG_URL$$', BLOG.get_site_base_url())

# Get the text for the last tweet from the configuration file. If the format is 
# not present, a default one is used.
# No variable can be used in the format
if conf.Summary.tweet_end: 
	if not conf.Summary.tweet_format_end_one:
		TWEET_END_ONE = "#blogReplay C'était le seul article publié cette semaine. Fin du spam! :)"
	else:
		TWEET_END_ONE = conf.Summary.tweet_format_end_one

	if not conf.Summary.tweet_format_end_many:
		TWEET_END_MANY = "#blogReplay C'était les %d articles publiés cette semaine. Fin du spam! :)"
	else:
		TWEET_END_MANY = conf.Summary.tweet_format_end_many


# After opening the log file, the tweets are sent.
with open(LOG_FILE, 'a') as log_file:
	log_file.write('[%s] %s'%(TODAY, base_dir) + "\n")
	if len(POSTS):		
		# Post the first tweet
		if conf.Summary.tweet_begin:
			first_tweet(TWEET_BEGIN)

		# Post a new tweet for each article
		for a in POSTS:
			time.sleep(interval)

			# Get the title and the URL of the article and, if necessary, 
			# shorten it.
			title = a['title']
			url = a['url']
			if BITLY_API:
				s = BITLY_API.shorten(url)
				u = s['url']
				url = unicodedata.normalize('NFKD', u).encode('ascii','ignore')
			

			# Calculate the maximum length for the title of the article.
			# This length is equal to : 
			# 	140 - length of TWEET_TEXT - spaces - length of the URL
			# 	140 is max length for a tweet 
			# If the title is too large to fit in a tweet, it is shorten
			# and some dots are added at its end.
			max_length = 140 - TWEET_FORMAT_LENGTH - 5 - len(url)
			if len(title) >= max_length:
				title = "%s%s"%(title[:max_length-3],"...")

			# Generate the tweet message from TWEET_FORMAT, and remove any extra
			# HTML character.
			tweet_text = TWEET_FORMAT.replace('$$POST_TITLE$$', title)
			tweet_text = tweet_text.replace('$$POST_URL$$', url)
			tweet_text = strip_tags(tweet_text)

			# Post the tweet on Twitter
			twitter_send(tweet_text)
			log_file.write(tweet_text + "\n")
			
		# At the end, publish the ending tweet.
		# If more than one tweet were sent, the last tweet will contains the number
		# of tweets. In this purpose, the TWEET_SUMMARY_END_MANY variable **has** 
		# to contains a '%d' (integer) replacement tag.
		if conf.Summary.tweet_end:
			if len(POSTS) == 1:
				twitter_send(TWEET_END_ONE)
				log_file.write(TWEET_END_ONE + "\n")
			else:	
				twitter_send(TWEET_END_MANY%len(POSTS))
				log_file.write(TWEET_END_MANY%len(POSTS) + "\n")
	else:
		# If no new articles were published, log it.
		log_file.write("No new posts\n")

# END
